File: Nimble.Testing/Testing.Services.Json/Scripts/dateinterval.js
var DateIntervalType = {
    Undefined:      { value: "Undefined",       code: 00,   description: "undefined"        },
    Today:          { value: "Today",           code: 01,   description: "today"            },
    Yesterday:      { value: "Yesterday",       code: 02,   description: "yesterday"        },
    CurrentWeek:    { value: "CurrentWeek",     code: 03,   description: "current week"     },
    LastWeek:       { value: "LastWeek",        code: 04,   description: "last week"        },
    CurrentMonth:   { value: "CurrentMonth",    code: 05,   description: "current month"    },
    LastMonth:      { value: "LastMonth",       code: 06,   description: "last month"       },
    CurrentQuarter: { value: "CurrentQuarter",  code: 07,   description: "current quarter"  },
    LastQuarter:    { value: "LastQuarter",     code: 08,   description: "last quarter"     },
    CurrentYear:    { value: "CurrentYear",     code: 09,   description: "current year"     },
    LastYear:       { value: "LastYear",        code: 10,   description: "last year"        }
};

var DayOfWeek = function () {
    var days = {
        Sunday:     { value: "Sunday",      code: 0 },
        Monday:     { value: "Monday",      code: 1 },
        Tuesday:    { value: "Tuesday",     code: 2 },
        Wednesday:  { value: "Wednesday",   code: 3 },
        Thursday:   { value: "Thursday",    code: 4 },
        Friday:     { value: "Friday",      code: 5 },
        Saturday:   { value: "Saturday",    code: 6 }
    };
    var first = days.Monday;
    var getFirst = function () {
        return first;
    };
    var setFirst = function (value) {
        for (var day in days) {
            if (value === days[day].value) {
                first = days[day];
                break;
            }
        }
    };
    var getWeekdays = function () {
        var weekdays = [];
        for (var day in days) {
            weekdays.push(days[day]);
        }
        return [].concat(weekdays.slice(first.code, weekdays.length)).concat(weekdays.slice(0, first.code));
    };
    var timeSpanDigits = function (value) {
        var digits = "";
        if (value > 9) {
            digits += value;
        } else {
            digits += "0" + value;
        }
        return digits;
    };
    var timeSpanPhrase = function (value, plural, singular) {
        var phrase = "";
        if (value > 0) {
            if (value > 1) {
                phrase += value + plural;
            } else {
                phrase += value + singular;
            }
        }
        return phrase;
    };
    var timeSpanRound = function (value) {
        if (value.seconds >= 30) {
            if (value.minutes === 59) {
                if (value.hours === 23) {
                    value.hours = 0;
                } else {
                    value.hours++;
                }
                value.minutes = 0;
            } else {
                value.minutes++;
            }
            value.seconds = 0;
        }
    };
    var getTimeSpan = function (seconds, format) {
        var timeSpan = {
            format: format,
            isPozitive: true,
            seconds: seconds,
            toString: ""
        };
        if (timeSpan.seconds != null) {
            if (timeSpan.seconds < 0) {
                timeSpan.isPozitive = false;
                timeSpan.seconds = Math.abs(timeSpan.seconds);
                timeSpan.toString += "-";
            }
            timeSpan.hours = Math.floor(timeSpan.seconds / 3600);
            timeSpan.seconds -= timeSpan.hours * 3600;
            timeSpan.minutes = Math.floor(timeSpan.seconds / 60);
            timeSpan.seconds -= timeSpan.minutes * 60;
            switch (timeSpan.format) {
                case 0:
                    {
                        timeSpan.toString +=
                            timeSpanDigits(timeSpan.hours) + ":" +
                            timeSpanDigits(timeSpan.minutes) + ":" +
                            timeSpanDigits(timeSpan.seconds);
                        break;
                    }
                case 1:
                    {
                        timeSpan.toString +=
                            timeSpanPhrase(timeSpan.hours, " hours ", " hour ") +
                            timeSpanPhrase(timeSpan.minutes, " minutes ", " minute ") +
                            timeSpanPhrase(timeSpan.seconds, " seconds ", " second ");
                        break;
                    }
                case 2:
                    {
                        timeSpanRound(timeSpan);
                        timeSpan.toString +=
                            timeSpanDigits(timeSpan.hours) + ":" +
                            timeSpanDigits(timeSpan.minutes);
                        break;
                    }
                case 3:
                    {
                        timeSpanRound(timeSpan);
                        timeSpan.toString +=
                            timeSpanPhrase(timeSpan.hours, " hours ", " hour ") +
                            timeSpanPhrase(timeSpan.minutes, " minutes ", " minute ");
                        break;
                    }
            }
        }
        return timeSpan;
    };
    return {
        days: days,
        getFirst: getFirst,
        setFirst: setFirst,
        getWeekdays: getWeekdays,
        getTimeSpan: getTimeSpan
    };
}();

function DateInterval() {
    this.dateNow = new Date();
    this.includeTime = false;
    this.dateIntervalType = DateIntervalType.Undefined;
    this.dateFrom = null;
    this.dateTo = null;
    this.getDateNow = function () {
        return this.dateNow;
    };
    this.getDateFrom = function () {
        var value = null;
        if (this.dateFrom != null) {
            if (this.includeTime) {
                value = new Date(this.dateFrom.getTime());
            } else {
                value = new Date(this.dateFrom.getFullYear(), this.dateFrom.getMonth(), this.dateFrom.getDate());
            }
        }
        return value;
    };
    this.getDateTo = function () {
        var value = null;
        if (this.dateTo != null) {
            if (this.includeTime) {
                value = new Date(this.dateTo.getTime());
            } else {
                value = new Date(this.dateTo.getFullYear(), this.dateTo.getMonth(), this.dateTo.getDate());
                value.setDate(value.getDate() + 1);
                value.setMilliseconds(value.getMilliseconds() - 3);
            }
        }
        return value;
    };
    this.setDate = function (date) {
        if (date == null) {
            this.dateFrom = null;
            this.dateTo = null;
        } else {
            date = date.getTime();
            this.dateFrom = new Date(date);
            this.dateTo = new Date(date);
        }
    };
    this.setWeek = function () {
        var days = DayOfWeek.getFirst().code - this.dateNow.getDay();
        this.dateFrom.setDate(this.dateFrom.getDate() + days);
        if (days > 0) {
            this.dateFrom.setDate(this.dateFrom.getDate() - 7);
        }
        this.dateTo.setDate(this.dateFrom.getDate() + 6);
    };
    this.setMonth = function () {
        this.dateTo.setMonth(this.dateTo.getMonth() + 1);
        this.dateTo.setDate(0);
    };
    this.setQuarter = function () {
        var months = (this.dateNow.getMonth() + 1) % 3;
        if (months === 0) {
            this.dateFrom.setMonth(this.dateFrom.getMonth() - 2);
        }
        else {
            this.dateFrom.setMonth(this.dateFrom.getMonth() + 1 - months);
            this.dateTo.setMonth(this.dateTo.getMonth() + 3 - months);
        }
        this.setMonth();
    };
    this.setYear = function () {
        this.dateFrom.setMonth(0);
        this.dateTo.setMonth(11);
        this.setMonth();
    };
    this.setDateInterval = function () {
        var date = new Date(this.dateNow.getTime());
        switch (this.dateIntervalType) {
            case DateIntervalType.Undefined:
                {
                    this.setDate(null);
                    break;
                }
            case DateIntervalType.Today:
                {
                    this.setDate(date);
                    break;
                }
            case DateIntervalType.Yesterday:
                {
                    date.setDate(date.getDate() - 1);
                    this.setDate(date);
                    break;
                }
            case DateIntervalType.CurrentWeek:
                {
                    this.setDate(date);
                    this.setWeek();
                    break;
                }
            case DateIntervalType.LastWeek:
                {
                    date.setDate(date.getDate() - 7);
                    this.setDate(date);
                    this.setWeek();
                    break;
                }
            case DateIntervalType.CurrentMonth:
                {
                    date.setDate(1);
                    this.setDate(date);
                    this.setMonth();
                    break;
                }
            case DateIntervalType.LastMonth:
                {
                    date.setDate(1);
                    date.setMonth(date.getMonth() - 1);
                    this.setDate(date);
                    this.setMonth();
                    break;
                }
            case DateIntervalType.CurrentQuarter:
                {
                    date.setDate(1);
                    this.setDate(date);
                    this.setQuarter();
                    break;
                }
            case DateIntervalType.LastQuarter:
                {
                    date.setDate(1);
                    date.setMonth(date.getMonth() - 3);
                    this.setDate(date);
                    this.setQuarter();
                    break;
                }
            case DateIntervalType.CurrentYear:
                {
                    date.setDate(1);
                    this.setDate(date);
                    this.setYear();
                    break;
                }
            case DateIntervalType.LastYear:
                {
                    date.setDate(1);
                    date.setYear(date.getFullYear() - 1);
                    this.setDate(date);
                    this.setYear();
                    break;
                }
        }
    };
}

File: Nimble.Business/Nimble.Business.Library/Common/DateInterval.cs
#region Using

using System;
using System.Runtime.Serialization;
using Nimble.Business.Library.Attributes;

#endregion Using

namespace Nimble.Business.Library.Common
{
    [DataContract]
    public enum DateIntervalType
    {
        [EnumMember]
        [FieldCategory(Name = "undefined")]
        Undefined,
        [EnumMember]
        [FieldCategory(Name = "today")]
        Today,
        [EnumMember]
        [FieldCategory(Name = "yesterday")]
        Yesterday,
        [EnumMember]
        [FieldCategory(Name = "current week")]
        CurrentWeek,
        [EnumMember]
        [FieldCategory(Name = "last week")]
        LastWeek,
        [EnumMember]
        [FieldCategory(Name = "current month")]
        CurrentMonth,
        [EnumMember]
        [FieldCategory(Name = "last month")]
        LastMonth,
        [EnumMember]
        [FieldCategory(Name = "current quarter")]
        CurrentQuarter,
        [EnumMember]
        [FieldCategory(Name = "last quarter")]
        LastQuarter,
        [EnumMember]
        [FieldCategory(Name = "current year")]
        CurrentYear,
        [EnumMember]
        [FieldCategory(Name = "last year")]
        LastYear
    }

    [DataContract]
    public class DateInterval
    {
        #region Private Members

        #region Properties

        private DateTimeOffset dateTimeOffsetNow = DateTimeOffset.Now;
        private DateTimeOffset? dateFrom;
        private DateTimeOffset? dateTo;

        #endregion Properties

        #endregion Private Members

        #region Public Members

        #region Properties

        [DataMember(EmitDefaultValue = false, Order = 0)]
        public DateTimeOffset DateTimeOffsetNow
        {
            get { return dateTimeOffsetNow; }
            set { dateTimeOffsetNow = value; }
        }

        [DataMember(EmitDefaultValue = false, Order = 1)]
        public bool IncludeTime { get; set; }

        [DataMember(EmitDefaultValue = false, Order = 2)]
        public DateIntervalType DateIntervalType { get; set; }

        [DataMember(EmitDefaultValue = false, Order = 3)]
        public DateTimeOffset? DateFrom
        {
            get
            {
                if (!dateFrom.HasValue ||
                    IncludeTime)
                {
                    return dateFrom;
                }
                return dateFrom.Value.DateOffset();
            }
            set
            {
                dateFrom = value;
            }
        }

        [DataMember(EmitDefaultValue = false, Order = 4)]
        public DateTimeOffset? DateTo
        {
            get
            {
                if (!dateTo.HasValue ||
                    IncludeTime)
                {
                    return dateTo;
                }
                return dateTo.Value.DateOffset().AddDays(1).AddMilliseconds(-3);
            }
            set
            {
                dateTo = value;
            }
        }

        #endregion Properties

        #region Methods

        public DateInterval(DateTimeOffset dateTimeOffsetNow, DateIntervalType dateIntervalType)
        {
            this.dateTimeOffsetNow = dateTimeOffsetNow;
            DateIntervalType = dateIntervalType;
            var quarterMonths = dateTimeOffsetNow.Month % 3;
            switch (DateIntervalType)
            {
                case DateIntervalType.Undefined:
                    {
                        dateFrom = dateTo = null;
                        break;
                    }
                case DateIntervalType.Today:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow;
                        break;
                    }
                case DateIntervalType.Yesterday:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow.AddDays(-1);
                        break;
                    }
                case DateIntervalType.CurrentWeek:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow;
                        if (dateTimeOffsetNow.DayOfWeek == DayOfWeek.Sunday)
                        {
                            dateFrom = dateFrom.Value.AddDays(-6);
                        }
                        else
                        {
                            dateFrom = dateFrom.Value.AddDays(1 - (int)dateFrom.Value.DayOfWeek);
                            dateTo = dateTo.Value.AddDays(7 - (int)dateTo.Value.DayOfWeek);
                        }
                        break;
                    }
                case DateIntervalType.LastWeek:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow.AddDays(-7);
                        if (dateTimeOffsetNow.DayOfWeek == DayOfWeek.Sunday)
                        {
                            dateFrom = dateFrom.Value.AddDays(-6);
                        }
                        else
                        {
                            dateFrom = dateFrom.Value.AddDays(1 - (int)dateFrom.Value.DayOfWeek);
                            dateTo = dateTo.Value.AddDays(7 - (int)dateTo.Value.DayOfWeek);
                        }
                        break;
                    }
                case DateIntervalType.CurrentMonth:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow;
                        dateFrom = new DateTimeOffset(
                            dateFrom.Value.Year, 
                            dateFrom.Value.Month, 
                            1,
                            dateFrom.Value.Hour,
                            dateFrom.Value.Minute,
                            dateFrom.Value.Second,
                            dateFrom.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = new DateTimeOffset(
                            dateTo.Value.Year,
                            dateTo.Value.Month,
                            1,
                            dateTo.Value.Hour,
                            dateTo.Value.Minute,
                            dateTo.Value.Second,
                            dateTo.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = dateTo.Value.AddMonths(1).AddDays(-1);
                        break;
                    }
                case DateIntervalType.LastMonth:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow.AddMonths(-1);
                        dateFrom = new DateTimeOffset(
                            dateFrom.Value.Year, 
                            dateFrom.Value.Month,
                            1,
                            dateFrom.Value.Hour,
                            dateFrom.Value.Minute,
                            dateFrom.Value.Second,
                            dateFrom.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = new DateTimeOffset(
                            dateTo.Value.Year, 
                            dateTo.Value.Month,
                            1,
                            dateTo.Value.Hour,
                            dateTo.Value.Minute,
                            dateTo.Value.Second,
                            dateTo.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = dateTo.Value.AddMonths(1).AddDays(-1);
                        break;
                    }
                case DateIntervalType.CurrentQuarter:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow;
                        if (quarterMonths == 0)
                        {
                            dateFrom = dateFrom.Value.AddMonths(-2);
                        }
                        else
                        {
                            dateFrom = dateFrom.Value.AddMonths(1 - quarterMonths);
                            dateTo = dateTo.Value.AddMonths(3 - quarterMonths);
                        }
                        dateFrom = new DateTimeOffset(
                            dateFrom.Value.Year, 
                            dateFrom.Value.Month,
                            1,
                            dateFrom.Value.Hour,
                            dateFrom.Value.Minute,
                            dateFrom.Value.Second,
                            dateFrom.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = new DateTimeOffset(
                            dateTo.Value.Year, 
                            dateTo.Value.Month,
                            1,
                            dateTo.Value.Hour,
                            dateTo.Value.Minute,
                            dateTo.Value.Second,
                            dateTo.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = dateTo.Value.AddMonths(1).AddDays(-1);
                        break;
                    }
                case DateIntervalType.LastQuarter:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow.AddMonths(-3);
                        if (quarterMonths == 0)
                        {
                            dateFrom = dateFrom.Value.AddMonths(-2);
                        }
                        else
                        {
                            dateFrom = dateFrom.Value.AddMonths(1 - quarterMonths);
                            dateTo = dateTo.Value.AddMonths(3 - quarterMonths);
                        }
                        dateFrom = new DateTimeOffset(
                            dateFrom.Value.Year, 
                            dateFrom.Value.Month,
                            1,
                            dateFrom.Value.Hour,
                            dateFrom.Value.Minute,
                            dateFrom.Value.Second,
                            dateFrom.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = new DateTimeOffset(
                            dateTo.Value.Year, 
                            dateTo.Value.Month,
                            1,
                            dateTo.Value.Hour,
                            dateTo.Value.Minute,
                            dateTo.Value.Second,
                            dateTo.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = dateTo.Value.AddMonths(1).AddDays(-1);
                        break;
                    }
                case DateIntervalType.CurrentYear:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow;
                        dateFrom = new DateTimeOffset(
                            dateFrom.Value.Year, 
                            1,
                            1,
                            dateFrom.Value.Hour,
                            dateFrom.Value.Minute,
                            dateFrom.Value.Second,
                            dateFrom.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = new DateTimeOffset(
                            dateTo.Value.Year, 
                            12,
                            31,
                            dateTo.Value.Hour,
                            dateTo.Value.Minute,
                            dateTo.Value.Second,
                            dateTo.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        break;
                    }
                case DateIntervalType.LastYear:
                    {
                        dateFrom = dateTo = dateTimeOffsetNow.AddYears(-1);
                        dateFrom = new DateTimeOffset(
                            dateFrom.Value.Year, 
                            1,
                            1,
                            dateFrom.Value.Hour,
                            dateFrom.Value.Minute,
                            dateFrom.Value.Second,
                            dateFrom.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        dateTo = new DateTimeOffset(
                            dateTo.Value.Year, 
                            12,
                            31,
                            dateTo.Value.Hour,
                            dateTo.Value.Minute,
                            dateTo.Value.Second,
                            dateTo.Value.Millisecond,
                            dateTimeOffsetNow.Offset);
                        break;
                    }
            }
        }

        public DateInterval(DateIntervalType dateIntervalType) : this(DateTimeOffset.Now, dateIntervalType)
        {
        }

        public DateInterval() : this(DateIntervalType.Undefined)
        {
        }

        #endregion Methods

        #endregion Public Members
    }
}